### ISSUE TITLE
$PATH keeps growing bigger

### ISSUE THREAD

**Issue by BarbzYHOOL** [fish-shell/fish-shell#5117]
fish 2.7

I've read the tutorial and the https://github.com/fish-shell/fish-shell/issues/527

It says to use:

    set PATH <mydir> $PATH

in the config.fish

but when doing so, the path keeps growing bigger with duplicate

I can add a "set -e" to remove duplicate in the conf, but it seems dirty.

Any way to permanently set PATH easily? (i'm trying to set the path to several machines at once, with automated install, so don't tell me "use universal variable" please :P)

**Comment by ridiculousfish** [comment]
The preferred ways to modify PATH are (here we'll add `/opt/local/bin`):

1. Add `set PATH /opt/local/bin $PATH` to `~/.config/fish/config.fish`, OR

2. Modify the fish_user_paths universal variable: `set fish_user_paths $fish_user_paths /opt/local/bin`. Note simply run this once, not in `config.fish`.

Can you please share the line you added and how you determined that path keeps growing bigger?

**Comment by mqudsi** [comment]
(btw, you can always use `uniq` to deduplicate a list:

```fish
set -x FOO one two two three
# now remove duplicate records:
set -x FOO (printf '%s\n' $FOO | sort -u)
# or with `uniq` instead of `sort -u`, which is faster but only handles duplicates that come after one another
set -x FOO (printf '%s\n' $FOO | uniq)
```

**Comment by raichoo** [comment]
I only set my `PATH` variable in a login shell (but on my setup I log in on the console and `startx` from there, I don't believe that many people prefer to work like that ^^). Most terminal emulators can be instructed start the shell as a login shell.

My `config.fish` looks something like this. I admit that this is a bit odd, but it works pretty well on my setup.

```
if status --is-login
  set -x PATH …
end
```

The `$fish_user_paths` doesn't do what I need due to other programs in my tool chain, so this is how I deal with that.

**Comment by ridiculousfish** [comment]
Thanks. What is in the '...' and how did you determine that PATH keeps growing? (i.e. what are steps to reproduce?)

**Comment by BarbzYHOOL** [comment]
Oops sorry, I forgot to answer (and read) I will answer asap, my bad

**Comment by mqudsi** [comment]
@BarbzYHOOL any updates or can we presume it's fixed?

**Comment by BarbzYHOOL** [comment]
I will retry tomorrow, I swear

**Comment by BarbzYHOOL** [comment]
1) Ok, so, again I retried simply as @ridiculousfish advised:

I add this to the config.fish:

    set PATH /tmp $PATH

Then I reload with my custom function OR with `exec fish` 3 times in a row:

    echo $PATH
    /tmp /tmp /tmp /usr/local/bin /usr/bin /bin /usr/games


2) The second method (`set fish_user_paths $fish_user_paths /opt/local/bin`), I can't use it because it's for automatic installation, it must be idempotent so I don't want to run a command to set my path.

3) The "uniq" method is a possible workaround but again, it seems dirty (unnecessary command at loading). Also I didn't understand the difference between sort -u and uniq

4) @raichoo method is doing nothing, it does not add anything to the path


So basically, I don't think it's fixed but it seems like there is nothing I can do but a workaround (like the "uniq" one)

**Comment by krader1961** [comment]
@BarbzYHOOL It's not as hard as it appears to get the behavior you want. It's hard to provide a single, always do it this way, answer because different people have different opinions and requirements.

The first thing to consider is that you may only want to modify `PATH` in login shells. That is, the shell your terminal emulator launches. You can do that by wrapping your modifications in a `if status is-login; ...; end` block. Then subshells, or your `exec fish` scenario, won't keep modifying the path var.

To keep from adding duplicates there is no need for `set -e PATH[$n]`. Which is good because that's hard to get right. All you have to do is check if the path is already present and add it only if it isn't present. For example,

```
contains /tmp $PATH
or set PATH /tmp $PATH
```

Not sure what @mqudsi had in mind. His comment about how to remove duplicate entries is fine when the order of the entries doesn't matter. But the order of the entries in `PATH` definitely matters so you should not use `sort -u` to eliminate duplicates. The difference between `sort -u` and `uniq` is that the latter does not change the order of the items but only removes duplicates adjacent to each other. In other words if you do `set VAR a a b c b d` doing `set VAR (printf '%s\n' $VAR | sort -u)` will remove the duplicates and is equivalent to `set VAR a b c d`. Using `uniq` instead would be equivalent to `set VAR a b c b d`. It doesn't remove the second `b` because it isn't adjacent to the first one -- it only removes the adjacent `a`.

Another option is to ignore the inherited `PATH` and always set it from scratch. This is what I do. So my *config.fish* looks like this:

```
# Customize the PATH. Start by exporting it with zero elements if we're not on Cygwin.
# If we're on Cygwin we'll want to keep its current contents.
if not string match -q 'CYGWIN*' $os_name
    set -gx PATH
end

test -d /usr/local/go/bin
and set -a PATH /usr/local/go/bin

test -d $HOME/go/bin
and set -a PATH $HOME/go/bin
```

This allows it to adapt to each system I run fish on. For example, I have the `go` language installed on my Mac's but none of my virtual machines that are running Linux, FreeBSD, etc. So I only add that directory to `PATH` if it exists.

**Comment by ridiculousfish** [comment]
Thanks for the update @BarbzYHOOL. It looks like your workflow is to reload fish from within fish. You'll have a similar issue with other shells, e.g. in bash:

```
bash-4.4$ cat ~/.bashrc
PATH=/foo:$PATH
bash-4.4$ exec bash
bash-4.4$ exec bash
bash-4.4$ exec bash
bash-4.4$ echo $PATH
/foo:/foo:/foo:/foo:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

The simplest thing to do is just use `contains` as krader suggested.

**Comment by BarbzYHOOL** [comment]
This works:

```contains /tmp $PATH
or set PATH /tmp $PATH```

This works too when I log in SSH, but since I'm using byobu (like screen/tmux), I don't understand how to update my path to existing sessions.

```if status is-login; ...; end ```

**Comment by mqudsi** [comment]
A very easy remedy for appending to `fish_user_paths` without running into this issue is to take advantage of variable shadowing: while `fish_user_paths` is a universal variable, you can set it without having its value balloon infinitely by using `set -g fish_user_paths foo $fish_user_paths` - the underlying `fish_user_paths` value won't be overridden (so this will never expand recursively) and `foo` gets appended to the path for the session.

**Comment by daydaygo** [comment]
i handle this in other way
```fish
# put this in the begin of config.fish
set fish_user_paths /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
set PATH /usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

**Comment by faho** [comment]
@daydaygo That just means fish_user_paths is useless, better just erase it.

**Comment by daydaygo** [comment]
> @daydaygo That just means fish_user_paths is useless, better just erase it.

get, only use PATH is just ok for me

**Comment by Pysis868** [comment]
Ran into that exec problem and found this thread.  Was getting around ~1000+ PATH entries......

A note for above, setting `fish_user_paths ` calls `__fish_reconstruct_path` which updates the PATH variable at least from the package config file: `/usr/local/Cellar/fish/*/share/fish/config.fish`
I have 3.0.2 currently going to 3.1.2 I think.

Made a small 'compacting' script I can run for my instance now.  As I start terminals I don't want to have that be a regular occurrence, and neither do I want to prepend every one of my PATH additive statements with contains, but might have to eventually..  Maybe a slightly better situation would to be set a sort of flag in the PATH variable or even another in the environment would be better, and skip large areas of variable-based config at least in this case since avoiding rephrasing them all as declarative statements.

**Comment by rayjanoka** [comment]
This helped me `set --erase --universal fish_user_paths`

**Comment by pantuza** [comment]
I have created the following fish function to Dedup $PATH variable entries while preserving order:

```fish
function dedup_path --description 'Remove duplicate entries from $PATH and preserves order'

    set --local new_path

    for path in $PATH
        if not contains $path $new_path
            set new_path $new_path $path
        end
    end

    set --global --export PATH $new_path
end
```

I have added `dedup_path` at the end of my `~/.config/fish/config.fish`
